### DISCUSSION TITLE
Do I need lerna to use auto for a monorepo?

### DISCUSSION THREAD

**Discussion by pomber** [intuit/auto#1861] [Q&A] [answered]
After reading the docs it isn't clear what happens when I run `auto shipit` on the root of a monorepo.

It does say it works with `lerna`:

> The npm plugin works out of the box with lerna in both independent and fixed mode. auto works on a repo basis and should be run from the root of the repo, not on each sub-package. No additional setup is required.

But is `lerna` required? 

(I'm currently using `yarn` as the 'installer' and `wsrun` as the "task-runner")

**Comment by hipstersmoothie** [comment, marked as answer]
At the moment yes it's required. In my opinion lerna is a pretty bad installer, an okay task runner, and a really good publisher. Auto uses lerna to do most of the tricky parts of publishing releases. 

@zephraph had an idea to make a tmp lerna.json to get around the user actually having to use lerna

**Reply by hipstersmoothie** [reply]
I've waffled back and forth over just reimplenting lerna's release logic in the npm plugin but the lerna team has done a lot of really great work there

**Reply by pomber** [reply]
thanks, is there any example repo using lerna and auto?

**Reply by just-be-dev** [reply]
Here's one (it uses independent versioning): https://github.com/zephraph/obsidian-tools/

It's relatively simple to setup, you can find the docs here: https://intuit.github.io/auto/docs/generated/npm#monorepo-usage

Note that if you're using an npm automation token to publish you're going to have to explicitly include this: 

https://github.com/zephraph/obsidian-tools/blob/814e11752b1a0eb86225dc7730b1d961acbaa5da/lerna.json#L5-L9

**Reply by hipstersmoothie** [reply]
Auto itself is also a good example

**Reply by pomber** [reply]
Thanks, I'm trying auto here: https://github.com/code-hike/codehike/pull/32

But canary versions are being skipped:
> ℹ  info      Skipping canary release due to PR being specifying no release. Use `auto canary --force` to override this setting

the PR has the `release` tag, what am I missing?

**Reply by hipstersmoothie** [reply]
The other label (`internal`) is a `none` release but that shouldn't skip the build. Taking a look now

**Reply by pomber** [reply]
the first try was with the "release" and without "internal", now I changed `onlyPublishWithReleaseLabel` to false and it worked

**Reply by hipstersmoothie** [reply]
I have a PR i'm about to merge to fix that behavior. Thanks for notifying me!

**Reply by hipstersmoothie** [reply]
https://github.com/intuit/auto/releases/tag/v10.18.5 should fix that. Sorry about the hiccup!

**Reply by pomber** [reply]
Awesome! I'll try it now, thank you.

Something else that isn't a big issue, the canary message is showing a private package `npm install storybook@MISSING` in the [comment](https://github.com/code-hike/codehike/pull/32#issue-591183971). Maybe it's my lerna config.

**Reply by hipstersmoothie** [reply]
That's fixed now too! https://github.com/intuit/auto/releases/tag/v10.18.7

**Reply by pomber** [reply]
Tried v10.18.7, but I think it's still trying to publish storybook, see: https://github.com/code-hike/codehike/pull/38

**Reply by hipstersmoothie** [reply]
Looking into it. That PR is odd, shouldn't have the changelog in it

**Reply by hipstersmoothie** [reply]
Ah my fix was only for version-less packages. since you added a version to the package it was included in the install list.

As for where there is a prerelease changelog in your PR I think it's because you're running `shipit` to many times per PR.

<img width="351" alt="Screen Shot 2021-03-14 at 11 40 32 AM" src="https://user-images.githubusercontent.com/1192452/111080103-18376880-84ba-11eb-8695-098af29f225c.png">

☝️ with that you'll get way to many releases happening:

Looks like 5 currently happen:

<img width="462" alt="Screen Shot 2021-03-14 at 11 41 38 AM" src="https://user-images.githubusercontent.com/1192452/111080129-4026cc00-84ba-11eb-8841-bf9f588f99c7.png">

**Reply by pomber** [reply]
Oh, thanks for looking into it. What triggers do you recommend for the GH Action?  
If I only have `push` I think it doesn't release the canary version because the PR doesn't exist at that point.

**Reply by hipstersmoothie** [reply]
I'll get a fix out for the first issue. For the multiple actions running I'll try to find the best setup for that too.

**Reply by hipstersmoothie** [reply]
It should still create the canary version, but if a PR hasn't been opened yet it has nowhere to post the version to. it will still be in the build logs though

**Reply by pomber** [reply]
but I'm using `onlyPublishWithReleaseLabel`

**Reply by hipstersmoothie** [reply]
`onlyPublishWithReleaseLabel` is only for your `latest` release line. Canaries are always published. I was talking with @zephraph about adding something like `onlyPublishCanaryWithReleaseLabel`

**Reply by hipstersmoothie** [reply]
Just verified that `push` will still publish a canary without an associated PR 

https://github.com/hipstersmoothie/auto-minimal-github-actions/runs/2107695038?check_suite_focus=true

**Reply by hipstersmoothie** [reply]
Fixed the private package inclusion. My github actions trigger recommendations are just `[push]`

**Reply by pomber** [reply]
Thanks, it makes sense. Maybe I can run a different workflow on [pr] to add the canary comment to the pr body.

**Reply by hipstersmoothie** [reply]
If you find a nice way to do that we can add it to the GitHub actions docs

**Reply by hydrosquall** [reply]
One compelling reason to explore ways to use auto in a monorepo without needing `lerna` is that `lerna` does not support yarn 2 `workspace:*` links. 

https://github.com/lerna/lerna/issues/2564

This makes it difficult to use `auto` in a yarn 2 monorepo where some packages depend on others in the same monorepo, given that the `workspace:*` format is the recommended way to specify a local dependency. (Otherwise, you run the risk of importing a remote rather than local version), and will need to manually update all dependents any time you update the version of a core package.

https://yarnpkg.com/features/workspaces#workspace-ranges-workspace

**Reply by hydrosquall** [reply]
Drilling into this a bit more, currently the main blocking command to trying to use `auto` in a yarn 2 monorepo with workspace links is here:

https://github.com/intuit/auto/blob/909ae8ba527d7805f71b3f065840ab2d37f798a1/packages/core/src/utils/get-lerna-packages.ts#L15

If this line were replaced with the option to use `yarn workspaces list` (or some yarn 2 equivalent with flags to get version numbers / private status / etc), it might be possible to use auto without lerna (although it's possible that some other commands would fail afterwards)

**Reply by hydrosquall** [reply]
Update: I was able to use `lerna` and `auto` with Yarn 2 by replacing all my `workspace:*` links with explicit version ranges (like `^0.2.0`)

**Comment by bguillen-tcs** [comment]
Hello @hipstersmoothie and @hydrosquall ,

I am new to `auto` and publishing packages, so forgive me if my question(s) are already answered above.  Things are just going over my head right now.  😂   Anyway, I am still a little confused about this.

Essentially, I've managed to setup a nice proof of concept utilizing `auto` to publish independent packages within a monorepo.  It truly works like a charm.  Feel free to check it out [here](https://github.com/bguillen-tcs/auto-poc)  -- the workflow and repo setup.

Now, after taking a step back with the team, I wanted to see if we could use `yarn@1.x`, as opposed to `lerna` for managing our monorepo packages -- essentially, publishing them independently, like how I am doing with `lerna`.  Is this possible?  To see what I am currently trying to do, checkout the repo [here](https://github.com/bguillen-tcs/auto-poc-yarn).  The main thing there is the release workflow; I am basically running `yarn install` from the root, compiling TS modules, and then running `auto shipit`.  However, it does not seem to publish properly -- the root `package.json` is getting the version bump, rather than actual package `package.json`.

I've taken a look through doc website, and have only come across monorepo documentation with `lerna`, [here](https://intuit.github.io/auto/docs/generated/npm#monorepo-usage).

Anyway, is publishing independent packages within a monorepo using `yarn` possible?  Is this something that is more of a limitation of `yarn` and not the `auto` software, or both?

Let me know if you need me to clarify anything.

All the best and thanks for this awesome tech, 

Bryan
